tests/rptest/tests/redpanda_test.py

```diff
@@ -306,3 +308,47 @@ def _consumer_offsets_present():
 
             # We are fully upgraded, yield to whatever the test wants to do in this version
             yield current_version
+
+
+class RedpandaMixedTest(RedpandaTestBase):
+    """
+    Base class for tests which can run either against vanilla infrastructure or
+    against the cloud.
+    """
+    def __init__(self,
+                 test_context: TestContext,
+                 *,
+                 min_brokers: int | None = None):
+        """
+        :param min_brokers: the minimum number of brokers the test requires. The
+        test must be OK with more brokers than the given number (which will occur
+        in cloud tests, as the number of brokers is already set at cloud cluster
+        creation time). None indicates that the test does not care how many brokers
+        are created (there will always be at least 1!).
+        """
+        super().__init__(test_context=test_context)
+        self.redpanda = make_redpanda_mixed_service(test_context,
+                                                    min_brokers=min_brokers)
+
+        self._client = DefaultClient(self.redpanda)
+
+    def setup(self):
+        super().setup()
+        if (cloud := self.as_cloud()):
+            wait_until(lambda: cloud.cluster_healthy(),
+                       timeout_sec=20,
+                       backoff_sec=5,
+                       err_msg='cluster unhealthy before start of test')
+
+    def as_cloud(self):
```

Comment:
Wouldn't you want the opposite helper method as well or is that a strict superset anyway?

Reply:
Yeah I can add it for symmetry. This utility method didn't end up being as useful as I'd thought because often you want if/else to cover both cases and then the utility methods don't really work, you want an `instance` check in the caller because then typing knows the type on both sides:

```
if (service := redpanda.as_cloud()):
  service has type RSC here
else:
  service = redpanda.as_vanilla();
  // unfortunately service has type RS | None here, not just RS, the type checker
  // isn't smart enough to intuit the type because it can't look inside as_cloud
```

Reply:
> is that a strict superset anyway?

It isn't, there are cloud-specific methods